curated_transformers/models/attention.py

```diff
@@ -61,12 +70,112 @@ def forward(
         attn_scores = q @ k.transpose(-2, -1)
         attn_scores /= math.sqrt(model_dim)
 
-        # Replace tokens that we don't want to attend to with a large
-        # negative value to zero them out during softmax normalization.
-        batch, seq_len = attn_mask.shape
-        attn_scores += attn_mask.logit_mask.view(batch, 1, 1, seq_len)
+        if attn_mask is not None:
+            # Replace tokens that we don't want to attend to with a large
+            # negative value to zero them out during softmax normalization.
+            batch, seq_len = attn_mask.shape
+            attn_scores += attn_mask.logit_mask.view(batch, 1, 1, seq_len)
 
         attn_weights = attn_scores.softmax(dim=-1)
         attn_values = self.dropout(attn_weights @ v)
 
         return attn_values
+
+
+# https://www.tensorflow.org/text/tutorials/transformer#multi-head_attention
+class SelfAttention(Module):
+    """Transformer self-attention layer (Vaswani et al., 2016)."""
+
+    def __init__(
+        self,
+        *,
+        dropout_prob: float = 0.1,
+        hidden_width: int = 768,
+        num_attention_heads: int = 12,
+    ):
+        """Construct a self-attention layer.
+
+        dropout_prob (float): dropout to apply between the self-attention
+            and output layers (default: 0.1).
+        hidden_width (int): hidden width of the layer (default: 768).
+        num_attention_heads (int): number of attenion heads (default: 12).
+        """
+        super().__init__()
+
+        self.model_dim = hidden_width
+        self.num_heads = num_attention_heads
+        if self.model_dim % self.num_heads != 0:
+            raise ValueError(
+                Errors.E003.format(
+                    hidden_width=self.model_dim, num_heads=self.num_heads
+                )
+            )
```

Comment:
Shall we inline this error here already?

Reply:
Inlined.